weave/monitoring2/openai.py

```diff
@@ -0,0 +1,401 @@
+
+        result = await self._base_create(*args, **kwargs)
+        self.context.outputs = result
+
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_end, self.context, *args, **kwargs)
+
+        return result
+
+    async def _streaming_create(self, *args, **kwargs):
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_send_request, self.context, *args, **kwargs)
+
+        stream = await self._base_create(*args, **kwargs)
+        async for chunk in stream:
+            for callback in self.callbacks:
+                await self._use_callback(callback.before_yield_chunk, self.context, *args, **kwargs)
+
+            yield chunk
+
+            for callback in self.callbacks:
+                await self._use_callback(callback.after_yield_chunk, self.context, *args, **kwargs)
+
+        for callback in self.callbacks:
+            await self._use_callback(callback.before_end, self.context, *args, **kwargs)
+
+    @staticmethod
+    async def _use_callback(f, context, *args, **kwargs):
+        if asyncio.iscoroutinefunction(f):
+            await f(context, *args, **kwargs)
+        else:
+            f(context, *args, **kwargs)
+
+
+class ChatCompletions:
+    def __init__(self, base_create, callbacks: List[Callback] = None):
+        self._base_create = base_create
+
+        self.callbacks = coalesce(callbacks, [])
+        self.callbacks += [
+            ReassembleStream(),
+            InfoMsg(),
+            LogToStreamTable.from_stream_name("test-stream3", "test-project", "megatruong"),
+        ]
+
+    def create(self, *args, **kwargs):
+        self.context = Context()
+        if kwargs.get("stream", False):
+            return self._streaming_create(*args, **kwargs)
+        return self._create(*args, **kwargs)
+
+    def _create(self, *args, **kwargs):
+        for callback in self.callbacks:
+            callback.before_send_request(self.context, *args, **kwargs)
+
+        result = self._base_create(*args, **kwargs)
+        self.context.outputs = result
+
+        for callback in self.callbacks:
+            callback.before_end(self.context, *args, **kwargs)
+
+        return result
+
+    def _streaming_create(self, *args, **kwargs):
+        for callback in self.callbacks:
+            callback.before_send_request(self.context, *args, **kwargs)
+
+        stream = self._base_create(*args, **kwargs)
+        self.context.chunks = []
+        for chunk in stream:
+            for callback in self.callbacks:
+                callback.before_yield_chunk(self.context, *args, **kwargs)
+
+            yield chunk
+            self.context.chunks.append(chunk)
+
+            for callback in self.callbacks:
+                callback.after_yield_chunk(self.context, *args, **kwargs)
+
+        for callback in self.callbacks:
+            callback.before_end(self.context, *args, **kwargs)
+
+
+def patch():
```

Comment:
No change explicitly requested, but might be worth looking at how we patch `wandb`: https://github.com/wandb/wandb/blob/main/wandb/integration/openai/openai.py